File: backend/services/smugmug.py
import time
import requests
from requests_oauthlib import OAuth1Session

SMUGMUG_BASE = "https://api.smugmug.com"
REQUEST_TOKEN_URL = "https://api.smugmug.com/services/oauth/1.0a/getRequestToken"
AUTHORIZE_URL = "https://api.smugmug.com/services/oauth/1.0a/authorize"
ACCESS_TOKEN_URL = "https://api.smugmug.com/services/oauth/1.0a/getAccessToken"

_RATE_DELAY = 0.2  # 5 req/sec max

def get_request_token(api_key: str, api_secret: str, callback_url: str):
    oauth = OAuth1Session(api_key, client_secret=api_secret, callback_uri=callback_url)
    r = oauth.fetch_request_token(REQUEST_TOKEN_URL)
    return r["oauth_token"], r["oauth_token_secret"]

def get_authorize_url(api_key: str, request_token: str) -> str:
    return f"{AUTHORIZE_URL}?oauth_token={request_token}&Access=Full&Permissions=Read"

def get_access_token(api_key: str, api_secret: str, request_token: str,
                     request_token_secret: str, verifier: str):
    oauth = OAuth1Session(
        api_key, client_secret=api_secret,
        resource_owner_key=request_token,
        resource_owner_secret=request_token_secret,
        verifier=verifier,
    )
    r = oauth.fetch_access_token(ACCESS_TOKEN_URL)
    return r["oauth_token"], r["oauth_token_secret"]

class SmugMugClient:
    def __init__(self, api_key: str, api_secret: str, token: str, token_secret: str):
        self.session = OAuth1Session(
            api_key, client_secret=api_secret,
            resource_owner_key=token,
            resource_owner_secret=token_secret,
        )

    def _get(self, path: str, params: dict = None):
        url = SMUGMUG_BASE + path
        p = {"_accept": "application/json", **(params or {})}
        time.sleep(_RATE_DELAY)
        r = self.session.get(url, params=p)
        r.raise_for_status()
        return r.json()["Response"]

    def get_user(self):
        data = self._get("/api/v2!authuser")
        return data["User"]

    def get_folders(self, user_nick: str, folder_path: str = "") -> list[dict]:
        """List subfolders under a path. folder_path="" = root."""
        path = f"/api/v2/folder/user/{user_nick}"
        if folder_path:
            path += f"/{folder_path.strip('/')}"
        path += "!folders"
        try:
            data = self._get(path)
            return data.get("Folder", [])
        except Exception:
            return []

    def get_folder_albums(self, user_nick: str, folder_path: str) -> list[dict]:
        """List albums directly inside a folder path."""
        path = f"/api/v2/folder/user/{user_nick}"
        if folder_path:
            path += f"/{folder_path.strip('/')}"
        path += "!albums"
        try:
            data = self._get(path)
            return data.get("Album", [])
        except Exception:
            return []

    def get_albums(self, user_nick: str):
        albums = []
        start = 1
        while True:
            data = self._get(f"/api/v2/user/{user_nick}!albums",
                             {"start": start, "count": 100})
            albums.extend(data.get("Album", []))
            pages = data.get("Pages", {})
            if pages.get("NextPage"):
                start += 100
            else:
                break
        return albums

    def get_albums_in_folder(self, user_nick: str, folder_path: str) -> list[dict]:
        """Recursively collect all albums under a folder path."""
        albums = self.get_folder_albums(user_nick, folder_path)
        for subfolder in self.get_folders(user_nick, folder_path):
            sub_path = subfolder.get("UrlPath", "").lstrip("/").split("/user/")[-1]
            # UrlPath is like /user/nick/2026 — extract the part after nick
            parts = sub_path.split("/", 1)
            rel = parts[1] if len(parts) > 1 else ""
            if rel:
                albums.extend(self.get_albums_in_folder(user_nick, rel))
        return albums

    def get_images(self, album_key: str):
        images = []
        start = 1
        while True:
            data = self._get(f"/api/v2/album/{album_key}!images",
                             {"start": start, "count": 100})
            images.extend(data.get("AlbumImage", []))
            pages = data.get("Pages", {})
            if pages.get("NextPage"):
                start += 100
            else:
                break
        return images

File: frontend/lib/smugmug.ts
/**
 * SmugMug OAuth 1.0a client — TypeScript port of backend/services/smugmug.py
 * Uses Node.js built-in crypto (no extra dependencies needed).
 */
import crypto from 'crypto'

const SMUGMUG_BASE      = 'https://api.smugmug.com'
const REQUEST_TOKEN_URL = `${SMUGMUG_BASE}/services/oauth/1.0a/getRequestToken`
const AUTHORIZE_URL     = `${SMUGMUG_BASE}/services/oauth/1.0a/authorize`
const ACCESS_TOKEN_URL  = `${SMUGMUG_BASE}/services/oauth/1.0a/getAccessToken`
const RATE_DELAY_MS     = 200

// ── OAuth 1.0a helpers ──────────────────────────────────────────────────────

function pct(s: string): string {
  return encodeURIComponent(s).replace(/[!'()*]/g, c => '%' + c.charCodeAt(0).toString(16).toUpperCase())
}

function sign(
  method: string,
  url: string,
  params: Record<string, string>,
  consumerSecret: string,
  tokenSecret = ''
): string {
  const normalized = Object.entries(params)
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
    .map(([k, v]) => `${pct(k)}=${pct(v)}`)
    .join('&')
  const base = `${method.toUpperCase()}&${pct(url)}&${pct(normalized)}`
  const key  = `${pct(consumerSecret)}&${pct(tokenSecret)}`
  return crypto.createHmac('sha1', key).update(base).digest('base64')
}

function oauthHeader(params: Record<string, string>): string {
  const parts = Object.entries(params)
    .filter(([k]) => k.startsWith('oauth_'))
    .map(([k, v]) => `${k}="${pct(v)}"`)
  return 'OAuth realm="", ' + parts.join(', ')
}

function baseParams(consumerKey: string, tokenKey = ''): Record<string, string> {
  const p: Record<string, string> = {
    oauth_consumer_key:     consumerKey,
    oauth_nonce:            crypto.randomBytes(16).toString('hex'),
    oauth_signature_method: 'HMAC-SHA1',
    oauth_timestamp:        Math.floor(Date.now() / 1000).toString(),
    oauth_version:          '1.0',
  }
  if (tokenKey) p.oauth_token = tokenKey
  return p
}

async function oauthPost(url: string, params: Record<string, string>): Promise<URLSearchParams> {
  const r = await fetch(url, {
    method: 'POST',
    headers: { Authorization: oauthHeader(params) },
  })
  if (!r.ok) throw new Error(`OAuth POST ${url} → ${r.status}: ${await r.text()}`)
  return new URLSearchParams(await r.text())
}

// ── Public auth functions ───────────────────────────────────────────────────

export async function getRequestToken(apiKey: string, apiSecret: string, callbackUrl: string) {
  const params: Record<string, string> = { ...baseParams(apiKey), oauth_callback: callbackUrl }
  params.oauth_signature = sign('POST', REQUEST_TOKEN_URL, params, apiSecret)
  const result = await oauthPost(REQUEST_TOKEN_URL, params)
  return {
    token:  result.get('oauth_token')!,
    secret: result.get('oauth_token_secret')!,
  }
}

export function getAuthorizeUrl(requestToken: string): string {
  return `${AUTHORIZE_URL}?oauth_token=${requestToken}&Access=Full&Permissions=Read`
}

export async function getAccessToken(
  apiKey: string,
  apiSecret: string,
  requestToken: string,
  requestSecret: string,
  verifier: string
) {
  const params: Record<string, string> = { ...baseParams(apiKey, requestToken), oauth_verifier: verifier }
  params.oauth_signature = sign('POST', ACCESS_TOKEN_URL, params, apiSecret, requestSecret)
  const result = await oauthPost(ACCESS_TOKEN_URL, params)
  return {
    token:  result.get('oauth_token')!,
    secret: result.get('oauth_token_secret')!,
  }
}

// ── SmugMug API client ──────────────────────────────────────────────────────

export class SmugMugClient {
  constructor(
    private apiKey: string,
    private apiSecret: string,
    private token: string,
    private tokenSecret: string
  ) {}

  async get(path: string, extraParams: Record<string, string> = {}): Promise<any> {
    await new Promise(r => setTimeout(r, RATE_DELAY_MS))
    const url     = SMUGMUG_BASE + path
    const qParams = { _accept: 'application/json', ...extraParams }
    const all: Record<string, string> = { ...baseParams(this.apiKey, this.token), ...qParams }
    all.oauth_signature = sign('GET', url, all, this.apiSecret, this.tokenSecret)
    const r = await fetch(`${url}?${new URLSearchParams(qParams)}`, {
      headers: { Authorization: oauthHeader(all) },
    })
    if (!r.ok) throw new Error(`SmugMug ${r.status}: ${path}`)
    return (await r.json()).Response
  }

  async getUser() {
    return (await this.get('/api/v2!authuser')).User
  }

  async getFolders(userNick: string, folderPath = ''): Promise<any[]> {
    let path = `/api/v2/folder/user/${userNick}`
    if (folderPath) path += `/${folderPath.replace(/^\//, '')}`
    try { return (await this.get(path + '!folders')).Folder ?? [] } catch { return [] }
  }

  async getFolderAlbums(userNick: string, folderPath = ''): Promise<any[]> {
    let path = `/api/v2/folder/user/${userNick}`
    if (folderPath) path += `/${folderPath.replace(/^\//, '')}`
    try { return (await this.get(path + '!albums')).Album ?? [] } catch { return [] }
  }

  async getAlbums(userNick: string): Promise<any[]> {
    const albums: any[] = []
    let start = 1
    while (true) {
      const data = await this.get(`/api/v2/user/${userNick}!albums`, { start: String(start), count: '100' })
      albums.push(...(data.Album ?? []))
      if (!data.Pages?.NextPage) break
      start += 100
    }
    return albums
  }

  async getImages(albumKey: string): Promise<any[]> {
    const images: any[] = []
    let start = 1
    while (true) {
      try {
        const data = await this.get(`/api/v2/album/${albumKey}!images`, { start: String(start), count: '100' })
        images.push(...(data.AlbumImage ?? []))
        if (!data.Pages?.NextPage) break
        start += 100
      } catch (e) {
        console.warn(`Skipping album ${albumKey}: ${e}`)
        break
      }
    }
    return images
  }
}
